src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +147,119 @@ internal void Freeze()
+        /// element type that's relevant), it is necessary for IL rewriters. Such rewriters also need to ensure
+        /// an appropriate alignment is maintained for the RVA field, and while they could also simplify by choosing
+        /// a worst-case alignment as does the compiler, they may instead use the .pack value as the alignment
+        /// to use for that field, since it's an opaque blob with no other indication as to what kind of data is
+        /// stored and what alignment might be required.
+        /// </param>
+        /// <returns>The field. This may have been newly created or may be an existing field previously created for the same data and alignment.</returns>
+        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort alignment)
         {
             Debug.Assert(!IsFrozen);
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
+
+            Cci.ITypeReference type = _proxyTypes.GetOrAdd(
+                ((uint)data.Length, Alignment: alignment), key =>
+                key.Size switch
+                {
+                    1 when _systemInt8Type is not null => _systemInt8Type,
+                    2 when _systemInt16Type is not null => _systemInt16Type,
+                    4 when _systemInt32Type is not null => _systemInt32Type,
+                    8 when _systemInt64Type is not null => _systemInt64Type,
+                    _ => new ExplicitSizeStruct(key.Size, key.Alignment, this, _systemValueType)
+                });
+
+            return _mappedFields.GetOrAdd((data, alignment), key =>
             {
-                var name = GenerateDataFieldName(data0);
-                var newField = new MappedField(name, this, type, data0);
-                return newField;
+                // For alignment of 1 (which is used in cases other than in fields for ReadOnlySpan<byte>),
+                // just use the hex value of the data hash.  For other alignments, tack on a '2', '4', or '8'
+                // accordingly.  As every byte will yield two chars, the odd number of chars used for 2/4/8
+                // alignments will never produce a name that conflicts with names for an alignment of 1.
+                Debug.Assert(alignment is 1 or 2 or 4 or 8, $"Unexpected alignment: {alignment}");
+                string hex = HashToHex(key.Data);
+                string name = alignment switch
+                {
+                    2 => $"{hex}2",
+                    4 => $"{hex}4",
+                    8 => $"{hex}8",
+                    _ => hex
+                };
+
+                return new MappedField(name, this, type, key.Data);
             });
         }
 
-        private Cci.ITypeReference GetStorageStruct(uint size)
+        private static string HashToHex(ImmutableArray<byte> data)
```

Comment:
>HashToHex [](http://example.com/codeflow?start=30&length=9)

Consider moving this function where ```GenerateDataFieldName``` used to be. That should make it easier to confirm that there is no unintentional behavior change in the code.

Reply:
I'll move it. Though fwiw is I changed its behavior a bunch of existing tests would be failing.

Reply:
> if I changed its behavior a bunch of existing tests would be failing

That is an optimistic assumption. Tests do not cover every possible situation. And that is one of  the reasons for having code reviews.

Reply:
Is it possible something could slip through? Of course. But especially with cryotigraphic hashes, and with a method that's just calling an existing hash function and converting the result to hex, it'd be a rare mistake that wouldn't fail every existing use of that hash.

But yes, code reviews are important.

Reply:
> But yes, code reviews are important.

And because of that, it helps to follow a general principle - if possible, do not move and modify code at the same time. Doing otherwise makes diff almost useless.

Reply:
I'm not going to spend more time debating this.